add guide for custom xcom backends (#196)
astronomer/airflow-guides: guides/custom-xcom-backends.md

---
title: "Custom XCom Backends"
description: "Creating a custom XCom backend with Airflow 2.0."
date: 2021-04-07T00:00:00.000Z
slug: "custom-xcom-backends"
heroImagePath: null
tags: ["Plugins", "XCom"]
---

## Overview

In December 2020, the Apache Airflow project made a huge step forward with the release of [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0). One of the main new features in this release is the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskflow-api), which was introduced to solve the challenge of explicitly passing messages between Airflow tasks. The TaskFlow API abstracts the task and dependency management layer away from users, and makes it much easier to work with XComs. 

> One of the features of the TaskFlow API that increases the flexibility of XComs is support for a [custom XCom backend](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=xcom#custom-xcom-backend). This means that rather than store XComs in Airflow's metadata database by default, you can push and pull XComs to and from an external system such as S3, GCS, or HDFS. You can also implement your own serialization / deserialization methods building off of the subclass of `BaseXCom`. 

Author: Yeah I see what you mean, I'll take it out here

## When to Use Custom XCom Backends

Before we dive into the details on *how* to set up an XCom backend, we want to briefly touch on *when* you might want to do so, and importantly, when you might not.

XCom backends provide much more flexibility than you would have using traditional XComs with Airflow's metadata database. The biggest practical differences are that when using a custom backend your XComs do not need to be Json serializable (because you can implement your own serialization/deserialization and store the data however you want), and the size of your XCom information is not limited by your metadata database. Additionally, using a custom backend would allow you to implement custom retention or backup policies, and can provide easier access to your XComs without having to deal with accessing the metadata database. This is more ideal for production environments, because you can more easily manage what happens to your XComs over time, and don't need to worry about periodically cleaning up the metadata database with another DAG.

> All of this can be great if you want to pass information between your tasks flexibly and in a sustainable manner. However, note that even though a custom XCom backend will allow you to pass more data between your tasks, Airflow is still not designed to be a processing framework. XComs were meant to pass *messages* between your tasks, e.g. metadata or *small* amounts of data, not large datasets. How much data you can safely pass in this case will be determined by how you have configured your XCom backend, as well as the infrastructure running your Airflow setup and the executor you have chosen. In general, if your data are large we recommend using an external processing framework such as Apache Spark, and using Airflow for orchestration. 

Author: Agree, makes sense

For more info on best practices when passing data between your tasks, check out our guide on that topic [here](https://www.astronomer.io/guides/airflow-passing-data-between-tasks).

### Example Use Case: Great Expectations

> As a more practical example, using [Great Expectations](https://greatexpectations.io/) with Airflow is an ideal use case for implementing a custom XCom backend. Great Expectations is an open source Python-based data validation framework, and thanks to the [Great Expectations Airflow provider](https://registry.astronomer.io/providers/great-expectations), it integrates seamlessly with Airflow.

Author: Good call, added a note in the section above.

These drawbacks can be easily solved by implementing a custom XCom backend which programmatically handles the results and saves them to an external file system. Below, we'll show a specific example of how you could set this up in an Astronomer project.

## Setting Up an XCom Backend

In this section, we'll cover the general steps for setting up an XCom backend with Airflow. Note that the steps here assume you are running Airflow in a Dockerized set-up. If you are using Astronomer, see the section below for detailed instructions on how to complete these steps on the platform.

### 1. Configure Your Backend

The first step to setting up a custom XCom backend is to configure the backend you want to use. The steps below apply to setting up an S3 backend on AWS, but the process would look similar if you used another service such as Azure blob storage, GCS, etc.

> - Create a bucket. In S3, make sure to block public access, enable versioning, and give the bucket a unique name.

Author: I put "your S3 account" - you could say "console", but this can also be done programmatically so kept it more general

- Create a policy for your bucket that will allow Airflow to access it. Ensure the policy allows read and write actions over the bucket as shown in the screenshot below.

> ![XCom Backend Policy](https://assets2.astronomer.io/main/guides/xcom/xcom_backend_policy.png)

Author: I had no idea how to do this, thank you!

- Create a user (e.g. "airflow-xcoms") with Programmatic access and attach the policy you created in Step 2 to that user. Note the access and secret keys for the steps in the next section.

### 2. Configure Your Airflow Environment

> The next step is configuring your Airflow environment to ensure it can connect to your custom backend. There are many ways you could accomplish this, some of which are more secure and sustainable than others. The end goal here is to define the relevant default connection such that it connects to your custom backend; for this example with S3, this means defining the `AIRFLOW_CONN_AWS_DEFAULT` environment variable, or setting up an `aws_default` connection in the UI. 

Author: Modified your suggestion a little bit because it doesn't have to be a 'default'; otherwise the update looks great.

3. Add your XCom backend python file with serialization/deserialization methods to the `include/` directory of your Astronomer project.
4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it should look like `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a couple of ways to add an environment variable to your Astronomer deployment, which are detailed [here](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables). In this case, we added the variable to our Dockerfile.
5. Deploy your project code to Astronomer, or start Airflow locally using the Astronomer CLI with `astro dev start`.
6. Add an Airflow connection to connect to your backend as shown in step 2 above. Note that if you are using a secrets backend with Astronomer, you can add the connection there.
7. Test your new XCom backend by running an appropriate DAG!

> For this example, we have created serialization and deserialization methods that take the results of the `GreatExpectationsOperator`, convert them to JSON, and save to a file on S3. Our custom backend class looks like this:

Author: Yes, they were meant to be different. The first deals with pandas dataframes generally, and this one is specifically to handle the great expectations use case. I wanted to show a general example, and then a more specific one, and include multiple serialization/deserialization methods to give people a sense of what you could do with them,

        task_id='ge_batch_kwargs_pass',
        expectation_suite_name='taxi.demo',
        batch_kwargs={
            'path': data_file,
            'datasource': 'data__dir'
        },
        data_context_root_dir=ge_root_dir,
    )

    # This runs an expectation suite against a data asset that passes the tests
    ge_batch_kwargs_list_pass = GreatExpectationsOperator(
        task_id='ge_batch_kwargs_list_pass',
        assets_to_validate=[
            {
                'batch_kwargs': {
                    'path': data_file,
                    'datasource': 'data__dir'
                },
                'expectation_suite_name': 'taxi.demo'
            }
        ],
        data_context_root_dir=ge_root_dir,
    )

    # This runs a checkpoint that will pass. Make sure the checkpoint yml file has the correct path to the data file.
    ge_checkpoint_pass = GreatExpectationsOperator(
        task_id='ge_checkpoint_pass',
        run_name='ge_airflow_run',
        checkpoint_name='taxi.pass.chk',
        data_context_root_dir=ge_root_dir,
    )

    
    ge_batch_kwargs_list_pass >> ge_batch_kwargs_pass >> ge_checkpoint_pass
```

> Now if we run this DAG, we will see three XCom files in our S3 bucket, one for each operator. With this implementation, we didn't need to enable XCom pickling, can version and access our XCom data easily, and are not filling up the Airflow metadata database, making this a much more sustainable way of using this popular operator.

Reviewer: ```suggestion
When we run this DAG, we will see three XCom files in our S3 bucket, one for each operator. With this implementation, we didn't need to enable XCom pickling, we can version and access our XCom data easily, and we are not filling up the Airflow metadata database, making this a much more sustainable way of using this popular operator.
```
I see the word "Sustainable" throughout this article, and it's not clear to me what it means in this context. Could "reliable" or "scalable" be a better choice here?

Author: In this context it means that you could keep using XComs in this way over time without any manual intervention (like to clean up your metadata database, which you would likely have to do with normal xcoms). I don't think "reliable" or "scalable" captures the same thought here, but maybe there's a better word?

## Using XCom Backends on the Astronomer Platform

Now that we've shown the general steps for setting up a custom XCom backend, in this section we'll show how to apply those steps when working with Airflow on the Astronomer platform. For this example, we'll use the Great Expectations use case for storing operator results XComs as described above.

> Note: All supporting code for this section can be found in [this repo](https://github.com/astronomer/custom-xcom-backend-tutorial).

First, use the following steps to configure your XCom backend on the Astronomer platform:

1. Configure your backend (S3, GCS, etc.) as described in Section 1 from the previous example.
2. Open an existing Astronomer project, or initialize a new one using `astro dev init` (if you don't already have the Astronomer CLI, install it using the [CLI Quickstart](https://www.astronomer.io/docs/cloud/stable/develop/cli-quickstart) before completing these steps).
3. Add the XCom backend Python file with serialization/deserialization methods as shown below to the `include/` directory of your Astronomer project.

> 4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it should look like `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer deployment, which are detailed in Astronomer's [Environment Variables guide](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables). In this case, we added the variable to our Dockerfile.

Reviewer: ```suggestion
4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it should look like `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer Deployment, which are detailed in Astronomer's [Environment Variables guide](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables). In this case, we added the variable to our Dockerfile.
```
